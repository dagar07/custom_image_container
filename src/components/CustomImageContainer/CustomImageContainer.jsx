import React, { Component } from 'react';
import PropTypes from 'prop-types';

class CustomImageContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      imgContainer: null
    };
    this.onImageLoader = this.onImageLoader.bind(this);
    this.onImageError = this.onImageError.bind(this);
    this.getImgDim = this.getImgDim.bind(this);
    this.createPlaceholderContainer = this.createPlaceholderContainer.bind(this);
    this.getPlaceholderStyle = this.getPlaceholderStyle.bind(this);
  }

  componentDidMount() {
    const { src, imgStyle } = this.props;

    let imgElement;
    this.setState((prevState, props) => {
      return {
        imgContainer: React.createElement('img', {
          src,
          onLoad: this.onImageLoader,
          onError: this.onImageError,
          style: {
            ...imgStyle,
            opacity: 0,
          },
          ref: ref => {
            imgElement = ref;
          }
        })
      };
    }, () => {
      this.getImgDim(imgElement);
    });
  }

  onImageError() {
    // on error show the alt tag
    // if you want to show other thing just added here
    this.onImageLoader();
  }

  onImageLoader() {
    // call on image load
    const {
      src, animation, animationDuration, easing, alt, imgStyle
    } = this.props;
    const imgBox = React.createElement('img', {
      src,
      alt,
      style: {
        ...imgStyle,
        animation: `${animation} ${animationDuration} ${easing}`,
        maxWidth: '100%',
        maxHeight: '100%',
      }
    });

    this.setState((prevState, props) => {
      return {
        imgContainer:  React.createElement(
          'div',
          {
            style: this.getPlaceholderStyle(imgStyle.width, imgStyle.height)
          },
          React.cloneElement(imgBox)
        )
      };
    });
  }

  createPlaceholderContainer(width, height) {
    const { imgContainer } = this.state;

    const placeholder = React.createElement(
      'div',
      {
        style: this.getPlaceholderStyle(width, height),
      },
      React.cloneElement(imgContainer)
    );

    this.setState((prevState) => {
      return {
        imgContainer: placeholder
      };
    });
  }

  getImgDim(imgElement) {
    const that = this;
    const dimInterval = setInterval(() => {
      const { width, height } = imgElement ? imgElement : {width : 0, height: 0};
      if (width && height) {
        clearInterval(dimInterval);
        that.createPlaceholderContainer(width, height);
      }
    }, 5);
  }

  getPlaceholderStyle(imgWidth, imgHeight) {
    // this function is used to set styling for placeholder
    const {
      placeholderStyle,
      placeholderImg,
      color,
    } = this.props;
    let placeholderStyling = {...placeholderStyle};

    if (color) {
      // if color attribute is defined then set it's initial one
      placeholderStyling = {
        ...placeholderStyling,
        backgroundColor: color
      }
    }

    if(typeof (placeholderImg) === 'string') {
      placeholderStyling = {
        ...placeholderStyling,
        backgroundImage: `url(${placeholderImg})`
      };
    }

    return {
      ...placeholderStyling,
      height: `${imgHeight}px`,
      width: `${imgWidth}px`
    };
  }

  render() {
    return this.state.imgContainer;
  }
}

CustomImageContainer.propTypes = {
  src: PropTypes.string.isRequired, // image soure value
  alt: PropTypes.string.isRequired, // this is alternative naame for umage
  imgStyle: PropTypes.object, // used to provide styling to img style tag
  placeholderImg: PropTypes.string, // used to provide placholder bg image
  placeholderStyle: PropTypes.object, // uset to set styling for placeholder
  color: PropTypes.string, // if only color property set then used for placeholder bg color
  animation: PropTypes.string, // used to show animation in image
  animationDuration: PropTypes.string, // animation duration
  easing: PropTypes.string, // animation function
};

CustomImageContainer.defaultProps = {
  placeholderImg: null,
  placeholderStyle: {
    position: 'relative',
    display: 'inline-block',
    backgroundPosition: 'center center',
    backgroundSize: 'cover',
    backgroundRepeat: 'no-repeat'
  },
  imgStyle: {},
  color: '#fefefe',
  animation: 'blurIn',
  animationDuration: '1s',
  easing: 'ease-in',
};

export default CustomImageContainer;