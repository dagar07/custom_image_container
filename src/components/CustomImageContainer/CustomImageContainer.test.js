import React from 'react';
import {configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import CustomImageContainer from './CustomImageContainer';
import renderer from 'react-test-renderer';

configure({ adapter: new Adapter() });
const src = 'https://image.shutterstock.com/image-photo/mountains-during-sunset-beautiful-natural-600w-407021107.jpg';
const alt = 'global Image';

describe('CustomImageContainer Component', () => {
    it('Snapshot test', () => {
        const tree = renderer
            .create(<CustomImageContainer src={src} alt={alt} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Check for "src" and "alt" prop to be present and to have a string value', () => {
        const el = shallow(<CustomImageContainer src={src} alt={alt}/>);

        expect(el.props().src).toEqual(src);
        expect(typeof el.props().src).toEqual('string');
    });

    it('Check for "placeholderImg" prop to be custom and to have a string value', () => {
        const placeholderImg = "http://getuikit.com/docs/images/placeholder_600x400.svg",
            el = mount(<CustomImageContainer src={src} alt={alt} placeholderImg={placeholderImg} />);

        expect(el.props().placeholderImg).toEqual(placeholderImg);
        expect(typeof el.props().placeholderImg).toEqual('string');
    });

    it('Check for "imgStyle" prop to have an object value', () => {
      const imgStyle = { backgroundColor: 'black' },
          el = mount(<CustomImageContainer src={src} alt={alt} imgStyle={imgStyle} />);

      expect(el.props().imgStyle).toEqual(imgStyle);
      expect(typeof el.props().imgStyle).toEqual('object');
    });

    it('Check for "placeholderStyle" prop to have an object value', () => {
        const placeholderStyle = { border: "2px solid red", backgroundColor: 'black' },
            el = mount(<CustomImageContainer src={src} alt={alt} placeholderStyle={placeholderStyle} />);

        expect(el.props().placeholderStyle).toEqual(placeholderStyle);
        expect(typeof el.props().placeholderStyle).toEqual('object');
    });

    it('Check for "animation" prop to have a string value', () => {
        const animation = 'bounceIn',
            el = mount(<CustomImageContainer src={src} alt={alt} animation={animation} />);

        expect(el.props().animation).toEqual(animation);
    });

    it('Check for "animationDuration" prop to have a string value', () => {
        const animationDuration = '1s',
            el = mount(<CustomImageContainer src={src} alt={alt} animationDuration={animationDuration} />);

        expect(el.props().animationDuration).toEqual(animationDuration);
        expect(typeof el.props().animationDuration).toEqual('string');
    });

    it('Check for "easing" prop to have a string value', () => {
        const easing = 'ease-in',
            el = mount(<CustomImageContainer src={src} alt={alt} easing={easing} />);

        expect(el.props().easing).toEqual(easing);
        expect(typeof el.props().easing).toEqual('string');
    });

    it('Check for rendered element to be an "img" with the expected src', () => {
        const el = mount(<CustomImageContainer src={src} alt={alt} />);

        expect(el.children().at(0).type()).toEqual('img');
        expect(el.find('img').prop('src')).toEqual(src);
    });
});