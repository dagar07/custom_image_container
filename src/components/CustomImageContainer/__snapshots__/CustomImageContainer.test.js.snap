// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomImageContainer Component Snapshot test 1`] = `
<img
  onError={[Function]}
  onLoad={[Function]}
  src="https://image.shutterstock.com/image-photo/mountains-during-sunset-beautiful-natural-600w-407021107.jpg"
  style={
    Object {
      "opacity": 0,
    }
  }
/>
`;
